File: ios-kyrie-ai/KyrieAI/Models/ScoreEngine.swift
//
//  ScoreEngine.swift
//  KyrieAI
//
//  Lightweight "AI" model that turns raw drill performance into category
//  scores and an overall Ball Handler Score.
//

import Foundation

nonisolated struct DrillSample {
    let move: Move
    /// 0...1 quality the camera/coach perceived for this rep set.
    let quality: Double
    /// reaction in ms (lower is better) — relevant for reaction moves.
    let reactionMs: Int
}

nonisolated enum ScoreEngine {
    /// Derive category scores (0...100) from assessment samples + the player's
    /// declared skill level baseline.
    static func categoryScores(from samples: [DrillSample], baseline: Int) -> [SkillCategory: Int] {
        var totals: [SkillCategory: Double] = [:]
        var counts: [SkillCategory: Double] = [:]

        for sample in samples {
            for skill in sample.move.primarySkills {
                totals[skill, default: 0] += sample.quality
                counts[skill, default: 0] += 1
            }
        }

        var result: [SkillCategory: Int] = [:]
        for category in SkillCategory.allCases {
            let avg = counts[category].map { totals[category]! / $0 } ?? 0.55
            // blend declared baseline with measured performance
            let measured = avg * 100
            let blended = Double(baseline) * 0.45 + measured * 0.55
            // small per-category variance so the radar looks alive
            let jitter = Double((category.rawValue.count * 7) % 9) - 4
            result[category] = max(20, min(99, Int(blended + jitter)))
        }
        return result
    }

    static func ballHandlerScore(from categories: [SkillCategory: Int]) -> Int {
        guard !categories.isEmpty else { return 0 }
        let weights: [SkillCategory: Double] = [
            .control: 1.3, .speed: 1.0, .coordination: 1.1,
            .reaction: 1.0, .creativity: 0.9, .weakHand: 1.05
        ]
        var weighted = 0.0
        var weightSum = 0.0
        for (cat, score) in categories {
            let w = weights[cat] ?? 1
            weighted += Double(score) * w
            weightSum += w
        }
        return Int((weighted / weightSum).rounded())
    }

    /// Convert a real Vision measurement of one rep window into a 0...1 quality.
    /// Blends how clearly the hands were tracked, how active the dribbling was,
    /// and how low/controlled the handle stayed, then weighs by difficulty.
    static func quality(from rep: HandleRepResult, difficulty: MoveDifficulty) -> Double {
        let detection = min(1, rep.detectionRate / 0.8)      // ~80% visibility = full marks
        let motion = min(1, rep.avgMotion / 0.05)            // active, committed dribbling
        let lowness = min(1, rep.avgLowness / 0.55)          // low, protected handle
        var quality = detection * 0.4 + motion * 0.35 + lowness * 0.25

        // Harder moves are graded a touch tougher.
        let penalty: Double
        switch difficulty {
        case .basic: penalty = 0
        case .intermediate: penalty = 0.04
        case .advanced: penalty = 0.08
        case .signature: penalty = 0.12
        }
        quality -= penalty
        return max(0.3, min(0.99, quality))
    }

    /// Score one completed live session into accuracy / reaction / xp.
    static func gradeSession(samples: [DrillSample], mode: TrainingMode) -> (accuracy: Int, avgReactionMs: Int, xp: Int) {
        guard !samples.isEmpty else { return (0, 0, 0) }
        let avgQuality = samples.map(\.quality).reduce(0, +) / Double(samples.count)
        let accuracy = max(35, min(99, Int(avgQuality * 100)))
        let avgReaction = samples.map(\.reactionMs).reduce(0, +) / samples.count
        let baseXP = mode.defaultReps * 10
        let bonus = Int(avgQuality * 80)
        return (accuracy, avgReaction, baseXP + bonus)
    }
}

File: android/app/src/main/java/com/rork/kyrieai/data/ScoreEngine.kt
package com.rork.kyrieai.data

import kotlin.math.roundToInt
import kotlin.random.Random

data class DrillSample(
    val move: Move,
    val quality: Double,
    val reactionMs: Int,
)

object ScoreEngine {
    /** Derive category scores (0..100) from assessment samples + declared baseline. */
    fun categoryScores(samples: List<DrillSample>, baseline: Int): Map<SkillCategory, Int> {
        val totals = mutableMapOf<SkillCategory, Double>()
        val counts = mutableMapOf<SkillCategory, Double>()
        for (sample in samples) {
            for (skill in sample.move.primarySkills) {
                totals[skill] = (totals[skill] ?: 0.0) + sample.quality
                counts[skill] = (counts[skill] ?: 0.0) + 1
            }
        }
        val result = mutableMapOf<SkillCategory, Int>()
        for (category in SkillCategory.entries) {
            val c = counts[category]
            val avg = if (c != null && c > 0) totals[category]!! / c else 0.55
            val measured = avg * 100
            val blended = baseline * 0.45 + measured * 0.55
            val jitter = ((category.label.length * 7) % 9) - 4
            result[category] = (blended + jitter).toInt().coerceIn(20, 99)
        }
        return result
    }

    fun ballHandlerScore(categories: Map<SkillCategory, Int>): Int {
        if (categories.isEmpty()) return 0
        val weights = mapOf(
            SkillCategory.CONTROL to 1.3, SkillCategory.SPEED to 1.0, SkillCategory.COORDINATION to 1.1,
            SkillCategory.REACTION to 1.0, SkillCategory.CREATIVITY to 0.9, SkillCategory.WEAK_HAND to 1.05,
        )
        var weighted = 0.0
        var weightSum = 0.0
        for ((cat, score) in categories) {
            val w = weights[cat] ?: 1.0
            weighted += score * w
            weightSum += w
        }
        return (weighted / weightSum).roundToInt()
    }

    /** A simulated quality estimate when no camera is available. */
    fun estimateQuality(difficulty: MoveDifficulty): Double {
        val base = Random.nextDouble(0.55, 0.92)
        val penalty = when (difficulty) {
            MoveDifficulty.BASIC -> 0.0
            MoveDifficulty.INTERMEDIATE -> 0.06
            MoveDifficulty.ADVANCED -> 0.12
            MoveDifficulty.SIGNATURE -> 0.18
        }
        return (base - penalty).coerceAtLeast(0.3)
    }

    data class SessionGrade(val accuracy: Int, val avgReactionMs: Int, val xp: Int)

    fun gradeSession(samples: List<DrillSample>, mode: TrainingMode): SessionGrade {
        if (samples.isEmpty()) return SessionGrade(0, 0, 0)
        val avgQuality = samples.sumOf { it.quality } / samples.size
        val accuracy = (avgQuality * 100).toInt().coerceIn(35, 99)
        val avgReaction = samples.sumOf { it.reactionMs } / samples.size
        val baseXP = mode.defaultReps * 10
        val bonus = (avgQuality * 80).toInt()
        return SessionGrade(accuracy, avgReaction, baseXP + bonus)
    }
}
